### PR TITLE
feat: Add `fmt.Stringer` to `unikraft.Nameable`

### LINKED ISSUE #268: Deprecate `unikraft.NameTypeVersion` in favour of `fmt.Stringer`
### Feature request summary

Implementations which utilize `unikraft.Nameable` should instead return an `fmt.Stringer` interface such that the canonical version of their name can be appropriately set and used across KraftKit.  This would then alleviate the use of the wrapper method [unikraft.TypeNameVersion](https://github.com/unikraft/kraftkit/blob/staging/unikraft/type.go#L111), which would return an incorrect value if wrapping a target.Target input which is unikraft.Nameable.

### Describe alternatives

_No response_

### Related architectures

None

### Related platforms

None

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by afzal442**
Closes: #268 

### Prerequisite checklist

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [ ] Tested your changes against relevant architectures and platforms;
  - [ ] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### CODE REVIEW

File: unikraft/type.go [hunk 1]
```diff
@@ -106,9 +109,9 @@ func PlaceComponent(workdir string, t ComponentType, name string) (string, error
 	return "", fmt.Errorf("cannot place component of unknown type")
 }
 
-// TypeNameVersion returns the canonical name of the component using the format
+// String() returns the canonical name of the component using the format
 // <TYPE>/<NAME>:<VERSION>
```
Comment (nderjung) [inline unikraft/type.go]:
```diff

- // TypeNameVersion returns the canonical name of the component using the format
- // String() returns the canonical name of the component using the format
- // <TYPE>/<NAME>:<VERSION>
+ // String implements fmt.Stringer and returns the canonical string
+ // representation of the nameable object
```

**Comment by afzal442** [comment]
Hi there, I hv been struggling wid some errs and can u take a look here? Do I have to create `struct` here? Thanks

```
Error: unikraft/type.go:114:7: invalid receiver type Nameable (pointer or interface type)
Error: unikraft/type.go:116:12: entity.Type undefined (type *Nameable is pointer to interface, not interface)
```

cc @nderjung

**Comment by nderjung** [comment]
Hi @afzal442, thanks for having a look at this! Interfaces in Go cannot be instantiated in the same way as structures and as such implementing the `String()` method on the interface directly will result in a compile-time error.  Instead, each structure which implements the interface must include the `String` method. 

My suggestion is to implement `String()` on all implementing components which in fact return the `unikraft.TypeNamVersion` such that their output is standardized, e.g. for the unikraft component:

```go
func (uc UnikraftConfig) String() string {
  return unikraft.TypeNameVersion(uc)
}
```

**Comment by afzal442** [comment]
Got tha idea. Thanks much! I have to implement String() on all implementing components like `UnikraftConfig`.

File: cmd/kraft/build/build.go [hunk 2]
```diff
@@ -129,7 +129,7 @@ func (opts *Build) Run(cmd *cobra.Command, args []string) error {
 		var packages []pack.Package
 		search := processtree.NewProcessTreeItem(
 			fmt.Sprintf("finding %s",
-				unikraft.TypeNameVersion(project.Template()),
+				project.Template().String(),
```
Comment (afzal442) [inline cmd/kraft/build/build.go:150]: Should I revert all of it?

File: unikraft/type.go [hunk 3]
```diff
@@ -106,9 +109,11 @@ func PlaceComponent(workdir string, t ComponentType, name string) (string, error
 	return "", fmt.Errorf("cannot place component of unknown type")
 }
 
-// TypeNameVersion returns the canonical name of the component using the format
+type MyNm string
+
+// String() returns the canonical name of the component using the format
 // <TYPE>/<NAME>:<VERSION>
-func TypeNameVersion(entity Nameable) string {
+func (entity MyNm) String() string {
```
Comment (afzal442) [inline unikraft/type.go]: Here too!

File: cmd/kraft/build/build.go (hunk 2 above)
Comment (nderjung) [reply cmd/kraft/build/build.go:150]: No, keep as is

File: unikraft/type.go (hunk 3 above)
Comment (nderjung) [reply unikraft/type.go]: No, [as per my previous comment](https://github.com/unikraft/kraftkit/pull/333#issuecomment-1509627769), please implement `String()` in each component and not in the interface.

**Review by nderjung — CHANGES_REQUESTED** [review]
Hi @afzal442 looks like you need to rebase to staging as it's including already committed code (in particular code from [this commit](https://github.com/unikraft/kraftkit/commit/5dbd551189e00aac8597cb2fb51a21f6032157d6) and [this commit](https://github.com/unikraft/kraftkit/commit/2bb730a0a7a300f2d4e570ef71d0595f4bb312b8)) which is resulting in merge conflicts (and so therefore the GitHub actions workflow cannot be run).

Please also correct the spelling mistakes in your commit message.

Thanks!

**Comment by nderjung** [comment]
Hi @afzal442, please can you view and address the checks that are failing? Additionally, rebasing to `staging` will invoke more up-to-date checks. Thanks!

**Comment by afzal442** [comment]
Sorry @nderjung ! I am done. Could you review it? Thanks

File: unikraft/type.go [hunk 4]
```diff
@@ -106,9 +109,9 @@ func PlaceComponent(workdir string, t ComponentType, name string) (string, error
 	return "", fmt.Errorf("cannot place component of unknown type")
 }
 
-// TypeNameVersion returns the canonical name of the component using the format
-// <TYPE>/<NAME>:<VERSION>
-func TypeNameVersion(entity Nameable) string {
+// String implements fmt.Stringer and returns the canonical string
+// representation of the nameable object
+func TypeNameVersion(entity fmt.Stringer) string {
```
Comment (nderjung) [inline unikraft/type.go]: You can discard this change.

**Comment by nderjung** [comment]
Please see my inline comment which requests to discard the highlighted change. Additionally, please squash your commits into a single commit which has a commit title the same as this PR. :)

**Comment by afzal442** [comment]
> Please see my inline comment which requests to discard the highlighted change

Hi @nderjung I had already pointed out [this](https://github.com/unikraft/kraftkit/pull/333#discussion_r1169237652) in fact. But you said no. That's why there was no point of deprecation of `unikraft.TypeNameVersion`. LMK if I understood here.

Edit: FYI, I am not able to squash since there are merge changes to this branch. If I try to squash all of it, a lot of commits will come up and end up with messing up internally. LMK if there is a way to fix it.

File: unikraft/type.go (hunk 4 above)
Comment (afzal442) [reply unikraft/type.go]: Yep! on it.

**Pull request closed without merging**